File: Módulo5 - WebServices/WS2/ws provider/provider.js
// server.js
const http = require('http');
const url = require('url');

// Função para validar CPF (mesma lógica da sua versão PHP/Python)
function validaCPF(cpf) {
  // remove caracteres que não são dígitos
  cpf = cpf.replace(/\D/g, '');

  if (cpf.length !== 11 || /^(\d)\1+$/.test(cpf)) {
    return false;
  }

  const calcDigito = (cpfSlice, pesoInicial) => {
    let soma = 0;
    for (let i = 0; i < cpfSlice.length; i++) {
      soma += parseInt(cpfSlice.charAt(i)) * (pesoInicial - i);
    }
    let resto = (soma * 10) % 11;
    return (resto === 10 ? 0 : resto);
  };

  const dig1 = calcDigito(cpf.slice(0, 9), 10);
  const dig2 = calcDigito(cpf.slice(0, 10), 11);

  return cpf === cpf.slice(0,9) + dig1.toString() + dig2.toString();
}

const server = http.createServer((req, res) => {
  const parsed = url.parse(req.url, true);
  const path = parsed.pathname;
  const method = req.method;

  // permitir resposta como JSON em todos os casos
  res.setHeader('Content-Type', 'application/json; charset=utf-8');

  if (path === '/cpf') {
    if (method === 'GET') {
      const query = parsed.query;
      const cpf = query.cpf || query.numero || '';
      const valid = validaCPF(cpf);
      res.statusCode = 200;
      res.end(JSON.stringify({ cpf: cpf, valid: valid }));
      return;
    }

    if (method === 'POST') {
      let body = '';
      req.on('data', chunk => {
        body += chunk;
      });
      req.on('end', () => {
        try {
          const data = JSON.parse(body);
          const cpf = data.cpf || '';
          const valid = validaCPF(cpf);
          res.statusCode = 200;
          res.end(JSON.stringify({ cpf: cpf, valid: valid }));
        } catch (err) {
          res.statusCode = 400;
          res.end(JSON.stringify({ error: 'JSON inválido' }));
        }
      });
      return;
    }
  }

  // Se rota não for /cpf ou método não suportado
  res.statusCode = 404;
  res.end(JSON.stringify({ error: 'endpoint not found' }));
});

const PORT = 8081;
server.listen(PORT, () => {
  console.log(`Servidor JS rodando em http://localhost:${PORT}`);
});

File: Módulo5 - WebServices/WS2/ws provider/provider.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json

def validar_cpf(cpf):
    cpf = ''.join(filter(str.isdigit, cpf))
    if len(cpf) != 11:
        return False
    if cpf == cpf[0] * 11:
        return False

    def calc_digito(cpf, peso):
        s = sum(int(cpf[i]) * (peso - i) for i in range(peso - 1))
        resto = (s * 10) % 11
        return 0 if resto == 10 else resto

    dig1 = calc_digito(cpf, 10)
    dig2 = calc_digito(cpf, 11)

    return cpf[-2:] == f"{dig1}{dig2}"


class CPFService(BaseHTTPRequestHandler):

    # ------------------------------
    # GET /cpf?cpf=XXXXXXXXXXX
    # ------------------------------
    def do_GET(self):
        from urllib.parse import urlparse, parse_qs
        parsed = urlparse(self.path)

        if parsed.path == "/cpf":
            params = parse_qs(parsed.query)
            cpf = params.get("cpf", [""])[0]

            valido = validar_cpf(cpf)
            self._send_json(200, {"cpf": cpf, "valid": valido})
            return

        self._send_json(404, {"error": "rota não encontrada"})

    # ------------------------------
    # POST /cpf
    # Body: { "cpf": "11144477735" }
    # ------------------------------
    def do_POST(self):
        if self.path == "/cpf":
            tamanho = int(self.headers.get('Content-Length', 0))
            corpo = self.rfile.read(tamanho).decode("utf-8")

            try:
                dados = json.loads(corpo)
                cpf = dados.get("cpf", "")
            except:
                self._send_json(400, {"error": "JSON inválido"})
                return

            valido = validar_cpf(cpf)
            self._send_json(200, {"cpf": cpf, "valid": valido})
            return

        self._send_json(404, {"error": "rota não encontrada"})

    # ------------------------------
    # Helper para JSON
    # ------------------------------
    def _send_json(self, status, data):
        resposta = json.dumps(data).encode()

        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(resposta)))
        self.end_headers()
        self.wfile.write(resposta)


def run():
    servidor = HTTPServer(("localhost", 8000), CPFService)
    print("Servidor rodando em http://localhost:8000 ...")
    servidor.serve_forever()


if __name__ == "__main__":
    run()
